src/lib/y2storage/dump_manager.rb

```diff
@@ -0,0 +1,327 @@
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "singleton"
+require "fileutils"
+require "yast"
+require "rspec/mocks"
+require "y2storage/devicegraph"
+require "y2storage/actions_presenter"
+require "y2storage/storage_manager"
+
+Yast.import "Mode"
+
+module Y2Storage
+  # Helper class to manage XML and YAML devicegraph dumps.
+  # Most of this is handling log rotation.
+  # This is a singleton class; use DumpManager.instance for all methods.
+  class DumpManager
+    include Singleton
+    include Yast::Logger
+
+    # The number of old dump directories to keep.
+    # This is in addition to the current one.
+    KEEP_OLD_DUMP_DIRS = 3
+
+    def initialize
+      @initialized = false
+    end
+
+    # Kill (recursively remove) all dump directories.
+    def kill_all_dump_dirs
+      Dir.glob(base_dir + "/storage*").each { |dir| FileUtils.remove_dir(dir) }
+    end
+
+    # Reset the dump directories: Rotate any old ones, create a new one and
+    # start numbering devicegraph dump files from zero.
+    def reset
+      rotate_dump_dirs
+      ensure_dump_dir
+      @prefix = "00"
+      @initialized = true
+    end
+
+    # Dump a devicegraph or the actions from an ActionsPresenter to file in
+    # human-readable format.
+    #
+    # If 'file_base_name' is specified, this is used. If it is nil, for
+    # devicegraphs an appropriate name based on its role (probed, staging) is
+    # generated, for actions, it is "actions".
+    #
+    # In any case, use a numbered prefix to make the sequence of files clear,
+    # and use the appropriate suffix (filename extension) depending on type.
+    #
+    # This will typically result in something like this:
+    #
+    #   01-probed.xml
+    #   01-probed.yml
+    #   02-staging.xml
+    #   02-staging.yml
+    #   03-staging.xml
+    #   03-staging.yml
+    #   04-actions.txt
+    #   05-committed.xml
+    #   05-committed.yml
+    #
+    # The directory to use is chosen according to the YaST mode (installation
+    # vs. installed system), and the directories are cleared (installation) or
+    # rotated (installed system) for each program invocation.
+    #
+    # @param dump_obj[Y2Storage::Devicegraph|Y2Storage::ActionsPresenter]
+    # @param file_base_name [String|nil] File base name to use.
+    #
+    # @return file base name with numeric prefix actually used
+    #   ("01-probed", "02-staging", ...)
+    #
+    def dump(dump_obj, file_base_name = nil)
+      return nil if dump_obj.nil?
+
+      if dump_obj.is_a?(Y2Storage::Devicegraph)
+        dump_devicegraph(dump_obj, file_base_name)
+      elsif dump_obj.is_a?(Y2Storage::ActionsPresenter)
+        dump_actions(dump_obj, file_base_name)
+      elsif dump_obj.is_a?(RSpec::Mocks::InstanceVerifyingDouble) ||
```

Comment:
Yes, you can add a `save` method for every time somebody uses `double` or `instance_double` in the unit tests. I started doing this, and it was a PITA. Moreover, we might choose to do one more thing when dumping a devicegraph, and then we'd need to add more methods for each one. And adding more things to `double` or `instance_double` everywhere typically means that rubocop will complain because one of its limits might easily be exceeded. Been there, started doing that, reverted it because it goes nowhere.

Remember that all this is code is supposed to help us debugging, not introduce new problems. The last thing we need is debugging code that makes maintaining our extensive test suites more awkward to use and maintain.

Reply:
BTW this is what we talked about in today's meeting.

Reply:
Ok, yes, but when you talk about it I was not sure what you exactly meant because I hadn't see the code, sorry.